### DISCUSSION TITLE
Mitigate connection errors

### DISCUSSION THREAD

**Discussion by joan1294** [sqlalchemy/sqlalchemy#11789] [Usage Questions] [answered]
I am getting some errors like the following:

```python
Traceback (most recent call last):
File "/tmp/8dcb7be882a5886/antenv/lib/python3.11/site-packages/sqlalchemy/pool/base.py", line 763, in _finalize_fairy
fairy._reset(pool, transaction_was_reset)
File "/tmp/8dcb7be882a5886/antenv/lib/python3.11/site-packages/sqlalchemy/pool/base.py", line 1038, in _reset
pool._dialect.do_rollback(self)
File "/tmp/8dcb7be882a5886/antenv/lib/python3.11/site-packages/sqlalchemy/dialects/mssql/base.py", line 2849, in do_rollback
super(MSDialect, self).do_rollback(dbapi_connection)
File "/tmp/8dcb7be882a5886/antenv/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 683, in do_rollback
dbapi_connection.rollback()
pyodbc.OperationalError: ('08S01', '[08S01] [Microsoft][ODBC Driver 17 for SQL Server]TCP Provider: Error code 0x68 (104) (SQLEndTran)')
```

My understanding of this error is that prior to this error there is a statement sent to the database which fails (please correct me if I am wrong here) and sqlalchemy goes ahead and rolls back the commit which triggers the error mentioned above which is a "Connection Reset by Peer error" (i.e. the server abruptly closed the connection) [link](https://techcommunity.microsoft.com/t5/azure-database-support-blog/lesson-learned-359-tcp-provider-error-code-0x68-104/ba-p/3834127).

I understand that the error is not comming from sqlalchemy but I have not been able to successfully mitigate this issue myself and I was looking for advise on what the best solution might be.

This error is happening in the context of an API using the fastAPI framework. This is what my engine settings look like:

```python
engine = create_engine(
        CNN_STR,
        future=True,
        implicit_returning=False,
        pool_pre_ping=True,
        pool_recycle=3600,
        fast_executemany=True
    )

def get_session() -> typing.Generator[Session, None, None]:
    with engine.connect() as conn:
        db = Session(bind=conn)
        try:
            yield db
        finally:
            db.close()
```

I thought that `pool_pre_ping=True` would help from what I read in [https://docs.sqlalchemy.org/en/14/core/pooling.html#dealing-with-disconnects](https://docs.sqlalchemy.org/en/14/core/pooling.html#dealing-with-disconnects) but I am still facing the error.

I was wondering if I should implement some kind of retry. The problem that I have with that is that I have with that solution is that I have a lot of queries and commits and I would like to be sure whether this is the best option and whether sqlalchemy provides a nice way to implement this via listeners, modifying the Session methods or something like that.

I would like to:

- Understand when is this issue happening. I don't get any other logs.
- How to mitigate the issue based on what I tried and the context explained above.

Many thanks

**Comment by zzzeek** [comment]
hi -

is that an error that is interrupting a request from completing or is it just a message in your logs that does not otherwise seem to impact anything?  the "finalize_fairy" method is used to dispose of database connections that were not explicitly closed by the application, if the stack trace begins on that method it would indicate that the application is allowing database connections to be de-referenced without the .close() method being called on them first (either directly or indirectly via other APIs which do this).

**Reply by joan1294** [reply]
Thank you for the reply,

At the beginning I thought it was interrupting my code but after investigating a bit I don't think it is. 
Regarding your point on the connection not being closed, is there a way to not get a connection closed if I use the following code to get a Session?

```python
def get_session() -> typing.Generator[Session, None, None]:
    with engine.connect() as conn:
        db = Session(bind=conn)
        try:
            yield db
        finally:
            db.close()
```
I would like to understand how could my code not close a connection given that I use this function to open a connection and get a session from it.

**Reply by zzzeek** [reply, marked as answer]
the code as written looks fine and the "with" block itself is enough to guarantee that the connection is correctly returned to the connection pool.

unfortunately when people are using FastAPI there seem to be conditions where the usual guarantees of synchronous Python code are lost, as it seems to make use of containers that are applying implicit greenlet/non-blocking IO to the program.

understanding your specific scenario would require information as to the error and complete stack trace which precedes the GC-related stack trace being shown

**Reply by joan1294** [reply]
Thank you for your feedback. That is enough for me to take it from here. I'll mark this as the answer.
